contracts/serverless-v2/RequestChainContract.sol

```diff
@@ -0,0 +1,317 @@
+        require(getRoleMemberCount(DEFAULT_ADMIN_ROLE) != 0, "AV:RR-All admins cant be removed");
+        return res;
+    }
+
+    function _authorizeUpgrade(
+        address /*account*/
+    ) internal view override onlyAdmin {}
+
+    //-------------------------------- Overrides end --------------------------------//
+
+    //-------------------------------- Initializer start --------------------------------//
+
+    function __RequestChainContract_init(
+        address _admin,
+        IERC20 _token,
+        EnclaveImage[] memory _images,
+        uint256 _globalMinTimeout,
+        uint256 _globalMaxTimeout,
+        uint256 _overallTimeout
+    ) public initializer {
+        __Context_init();
+        __ERC165_init();
+        __AccessControlEnumerable_init();
+        __AttestationAuther_init_unchained(_images);
+        __UUPSUpgradeable_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, _admin);
+
+        require(address(_token) != address(0), "INVALID_TOKEN");
+        token = _token;
+
+        require(_globalMinTimeout < _globalMaxTimeout, "INVALID_GLOBAL_TIMEOUTS");
+        globalMinTimeout = _globalMinTimeout;
+        globalMaxTimeout = _globalMaxTimeout;
+        overallTimeout = _overallTimeout;
+    }
+
+    //-------------------------------- Initializer end --------------------------------//
+
+    //-------------------------------- Gateway start --------------------------------//
+
+    IERC20 public token;
+
+    uint256 public globalMinTimeout;
+    uint256 public globalMaxTimeout;
+    uint256 public overallTimeout;
+
+    struct Gateway {
+        address operator;
+        bool status;
+    }
+
+    // enclaveKey => Gateway
+    mapping(address => Gateway) public gateways;
+
+    modifier onlyGatewayOperator(bytes memory _enclavePubKey) {
+        address enclaveKey = _pubKeyToAddress(_enclavePubKey);
+        require(
+            gateways[enclaveKey].operator == _msgSender(),
+            "ONLY_GATEWAY_OPERATOR"
+        );
+        _;
+    }
+
+    event GatewayRegistered(
+        bytes enclavePubKey,
+        address indexed enclaveAddress,
+        address indexed operator
+    );
+
+    event GatewayDeregistered(bytes enclavePubKey);
+
+    function registerGateway(
+        bytes memory _attestation,
+        bytes memory _enclavePubKey,
+        bytes memory _PCR0,
+        bytes memory _PCR1,
+        bytes memory _PCR2,
+        uint256 _timestampInMilliseconds
+    ) external {
+        // attestation verification
+        // bytes32 imageId = keccak256(abi.encodePacked(_PCR0, _PCR1, _PCR2));
+        // _verifyKey(
+        //     _attestation, 
+        //     _enclavePubKey, 
+        //     imageId, 
+        //     _enclaveCPUs, 
+        //     _enclaveMemory, 
+        //     _timestampInMilliseconds
+        // );
+        _verifyEnclaveKey(_attestation, IAttestationVerifier.Attestation(_enclavePubKey, _PCR0, _PCR1, _PCR2, _timestampInMilliseconds));
```

Comment:
_verifyEnclaveKey() returns false only when the enclaveKey has already been whitelisted. In case of invalid attestation, the function reverts. So we aren't checking the return bool value.